common/lib/core-interfaces/src/handles.ts

```diff
@@ -39,6 +39,12 @@ export interface IFluidHandleContext extends IProvideFluidHandleContext {
     attachGraph(): void;
 
     resolveHandle(request: IRequest): Promise<IResponse>;
+
+    /**
+     * Called when a handle belonging to this context is decoded.
+     * @param url - The url of the decoded handle.
+     */
+    handleDecoded(url: string): void;
```

Comment:
This object for most part had nothing to do with GC till now. It was used for creation of paths and resolving URIs.
Is there a way to keep GC & routing as a separate concerns?
They overlap for sure (and former depends on latter), but can latter have only one responsibility and not get tied directly into GC interactions?

I believe what we deal with here is a contract - a requirement to IFluidDataStore to do something (track new paths) and expose it to GC at a later time. If so, should it be explicit requirement (API) on IFluidDataStoreChannel, and not this hidden and very implicit requirement to call this API?
I.e. if I'm a developer building implementation of IFluidDataStoreChannel, how would I know that I need to call it? when?

Reply:
I do not like this approach as well. I proposed something along the lines of what are you saying here - https://github.com/microsoft/FluidFramework/pull/8053#issuecomment-954991925

I think we can extend the API surface of the IFluidSerializer by adding two more - getEncodedHandles() and getDecodedHandles() which would returns all handles serialized and parsed during its lifetime. We can then use this information accordingly for GC.

Reply:
I think serializer can continue to be single purpose object, but it could take a callback on creation to notify someone about handles (or have eventing mechanism for that). Yes, this seems to be a bit better, but please take a look at comments about being more explicit - how do we ensure that developer who decides to reimplement most of these interfaces knows what he/she needs to do? I.e. how does system either forces developer (through compilation errors, assuming usage of TS) to do right things, or we have very clear documentation and protection in the code that very quickly tells developers they are missing something?

Reply:
I am not sure how the class taking a callback solves this problem. We can add documentation and comments but it does not define any contract right. Basically, since it's not built into the IFluidSerializer interface, it does not force developers to implement this functionality. We will need to rely on documentation and example implementations to do this.